mesonbuild/compilers/c.py

```diff
@@ -168,21 +168,21 @@ def get_include_args(self, path, is_system):
     def get_std_shared_lib_link_args(self):
         return ['-shared']
 
-    def get_library_dirs_real(self):
-        env = os.environ.copy()
-        env['LC_ALL'] = 'C'
-        stdo = Popen_safe(self.exelist + ['--print-search-dirs'], env=env)[1]
-        paths = []
-        for line in stdo.split('\n'):
-            if line.startswith('libraries:'):
-                libstr = line.split('=', 1)[1]
-                paths = [os.path.realpath(p) for p in libstr.split(':') if os.path.exists(os.path.realpath(p))]
-        return paths
-
-    def get_library_dirs(self):
-        key = tuple(self.exelist)
+    def get_library_dirs_real(self, env):
+        code=''
+        extra_args = ['--print-search-dirs']
+        with self._build_wrapper(code, env, extra_args, None, 'compile', True) as p:
+            paths = []
+            for line in p.stdo.split('\n'):
+                if line.startswith('libraries:'):
+                    libstr = line.split('=', 1)[1]
+                    paths = [os.path.realpath(p) for p in libstr.split(':') if os.path.exists(os.path.realpath(p))]
+            return paths
+
+    def get_library_dirs(self, env=None):
```

Comment:
Is `get_library_dirs()` called from anywhere that doesn't have access to the environment? If no, we should make this a required argument so we don't have hidden bugs. Then you can also remove the conditional below in `_get_compiler_check_args()`.

Reply:
It is generally available, except in the case for mesonbuild/backend/backends.py, where we are filtering out rpath dirs which are in the system directory path. There _libdir_is_system calls get_library_dirs() without access to any environment. [That in turn is called by rpaths_for_bundled_shared_libraries which similarly doesn't have any env info]

As an aside: Is there a reason why the cross build compiler flags are stored in an environment object, rather than in a compiler object for the cross-compiler in question? I would have thought that for cross builds we would have multiple compiler objects, each tracking it's own args.

Reply:
> As an aside: Is there a reason why the cross build compiler flags are stored in an environment object, rather than in a compiler object for the cross-compiler in question?

Just cruft from ages past. I think someone recently opened a PR to change this.

Reply:
> mesonbuild/backend/backends.py

The `Backend` object has the environment object available as `self.environment`, so you should pass that.